4.x controller factory

ADmad opened the pull request:
Alternative to #13826 

### Todo
- [x]  ~~Move action invocation code under `ControllerFactory::invoke()` and remove `Controller::invokeAction()`.~~  I added a `Controller::getAction()` instead.
- [x]  Move `Http\ControllerFactory` to under `Controller`?

ADmad committed:
Move controller action invocation into controller factory.

ADmad committed:
Make BaseApplication directly use ControllerFactory.

This makes ActionDispatcher redundant and hence it's removed.

ADmad committed:
Remove no longer needed ControllerInterface.

ADmad committed:
Add ControllerFactoryInterface

ADmad committed:
Add Controller::getAction() which returns closure for action.

This will in future allow the controller factor to introspect the
action arguments and inject values fetched from DI container.

ADmad committed:
Move Http\ControllerFactory to Controller\ControllerFactory.

ADmad committed:
Fix annotation

ADmad committed:
Add missing dependency

ADmad committed:
Add missing "use" statement in readme example

ADmad committed:
Add missing dependency

ADmad committed:
Don't create ControllerFactory instance until needed.

markstory commented on src/Controller/Controller.php:
```diff
@@ -509,10 +508,20 @@ public function invokeAction(): ?ResponseInterface
 
-        /** @var callable $callable */
-        $callable = [$this, $request->getParam('action')];
+        return Closure::fromCallable([$this, $action]);
```
Does this allow Reflection to introspect the parameters of the action?

ADmad replied:
`ReflectionFunction` does allow introspecting closure args, I'll add a test for it

markstory commented on src/Controller/Controller.php:
```diff
@@ -509,10 +508,20 @@ public function invokeAction(): ?ResponseInterface
+     */
+    public function invokeAction(Closure $action, array $args): void
+    {
+        $result = $action(...$args);
```
Are you thinking that a DI container could be another closure that wraps the action's closure and performs injection?

ADmad replied:
I was going by what's done here https://github.com/robotusers/cakephp-di/blob/master/src/Http/ControllerFactory.php#L106-L134. It manually goes through the arguments list of the closure and generates list of args to pass.

Happy to change if there's a better way to invoke the closure with required args.

markstory replied:
This will work well. Building a new argument list with the injected parameters is simpler than what I was thinking.

markstory commented on src/Controller/ControllerFactory.php:
```diff
@@ -14,27 +14,33 @@
  * @license       https://opensource.org/licenses/mit-license.php MIT License
  */
-namespace Cake\Http;
+namespace Cake\Controller;
```
:+1: to the move. Can we include a `class_alias` and deprecation stub for the old name?

ADmad replied:
Done

ADmad committed:
Add test.

The test shows that closure returned by getAction() allows
intropecting the action's args.

ADmad committed:
Add deprecated class stub

ADmad committed:
Update typehints

ADmad committed:
Remove unneeded argument.

Controller can create it's own reponse instance.

markstory commented on src/Controller/ControllerFactory.php:
```diff
@@ -0,0 +1,149 @@
+        }
+
+        /** @var \Cake\Controller\Controller $controller */
+        $controller = $reflection->newInstance($request);
```
This is a nice simplification. The controller can create its own response as needed.

markstory commented on src/Http/composer.json:
```diff
@@ -26,7 +26,9 @@
-        "cakephp/utility": "^4.0"
+        "cakephp/utility": "^4.0",
+        "psr/http-server-handler": "^1.0",
+        "zendframework/zend-diactoros": "^2.1"
```
:+1:
